### DISCUSSION TITLE
Rich Presence functions for redundant behavior

### DISCUSSION THREAD

**Discussion by SporyTike** [RetroAchievements/RAIntegration#888] [Ideas]
When creating rich presences for games it sometimes happens that you need to use conditional lines for specific output to match the players progress or to avoid early spoilers. Normally this is not an issue because conditional outputs are fairly easy made but if they are connected with huge lookups or formats due to calculations or pointer chains it becomes pretty character consuming (e.g. glossary completion of an RPG, custom names with 5+ characters, ...). Having the same lookup/format chain for each conditional output makes the RP unnecessary longer, harder to read and more time consuming when you need to do a change in it because you need to apply it on all use-cases.

As example this example RP which checks for either one or two characters with custom name plus some collectible calculation:
```
Lookup:name
0x01=A
0x02=B
0x03=C

Display:
?0xH23502=3?In the Titlescreen
?0xH23502=4?In the File Select
?0xH23507=2?@name(0xH23510)@name(0xH23511)@name(0xH23512)@name(0xH23513) and @name(0xH23514)@name(0xH23515)@name(0xH23516)@name(0xH23517) are in @location(0xH23520) [Books: @Number(0xK23522_0xK23523_0xT23523*-1_0xS23523*-1)]
@name(0xH23510)@name(0xH23511)@name(0xH23512)@name(0xH23513) is in @location(0xH23520) [Books: @Number(0xK23522_0xK23523_0xT23523*-1_0xS23523*-1)]
```

My suggestion would be additional functions you can create in which you already setup what happens when this function is called:
```
Lookup:name
0x01=A
0x02=B
0x03=C

Function:books
@Number(0xK23522_0xK23523_0xT23523*-1_0xS23523*-1)

Function:nameCharA
@name(0xH23510)@name(0xH23511)@name(0xH23512)@name(0xH23513)

Function:nameCharB
@name(0xH23514)@name(0xH23515)@name(0xH23516)@name(0xH23517)

Display:
?0xH23502=3?In the Titlescreen
?0xH23502=4?In the File Select
?0xH23507=2?{nameCharA} and {nameCharB} are in @location(0xH23520) [Books: {books}]
{nameCharA} is in @location(0xH23520) [Books: {books}]
```

The length of the display lines can become significantly shorter and when something needs an edit - for example the calculation of the books needs to be adjusted - the dev can easily do it in the function itself and it's corrected everywhere in the display.

In this small example RP it might not be that obvious how useful it can be, but when you have calculations that are hundrets of characters long or way more conditional checks than it could become pretty handy.

**Comment by Jamiras** [comment]
I like the idea, but I'd use the same syntax as the other macros:
```
?0xH23507=2?@nameCharA() and @nameCharB() are in @location(0xH23520) [Books: @books()]
@nameCharA() is in @location(0xH23520) [Books: @books()]
```

I'd probably also rename it to "String" or "Text" or something similar as it's effectively a subset of a `Display` string. A "Function" implies the possibility of logic.

**Reply by SporyTike** [reply]
Sure, how it's implemented is something you can likely better decide than me, I just wanted to bring up the idea. I only used this syntax to make it clear what I mean and to avoid it being somehow confused with the ones we already have.
